python/ray/tests/test_worker_state.py

```diff
@@ -0,0 +1,162 @@
+
+                # Pause 5 seconds inside debugger
+                with ray._private.worker.global_worker.worker_paused_by_debugger():
+                    time.sleep(5)
+
+    def verify(num_paused):
+        workers = list_workers(
+            filters=[("num_paused_tasks", "=", num_paused)], detail=True
+        )
+        if len(workers) == 0:
+            return False
+        worker = workers[0]
+        return worker.num_paused_tasks == num_paused
+
+    test_actor = TestActor.options(max_concurrency=actor_concurrency).remote()
+    refs = [  # noqa
+        test_actor.main_task.options(name=f"TestActor.main_task_{i}").remote(i)
+        for i in range(20)
+    ]
+
+    wait_for_condition(verify, num_paused=1)
+
+
+@pytest.mark.parametrize("actor_concurrency", [1, 3])
+def test_worker_paused_threaded_actor(shutdown_only, actor_concurrency):
+    ray.init(_system_config=_SYSTEM_CONFIG)
+
+    @ray.remote
+    class ThreadedActor:
+        def main_task(self, i):
+            def thd_task():
+                if i == 0:
+                    import time
+
+                    # Pause 5 seconds inside debugger
+                    with ray._private.worker.global_worker.worker_paused_by_debugger():  # noqa: E501
+                        time.sleep(5)
+
+            thd = threading.Thread(target=thd_task)
+            thd.start()
+            thd.join()
+
+    def verify(num_paused):
+        workers = list_workers(
+            filters=[("num_paused_tasks", "=", num_paused)], detail=True
+        )
+        if len(workers) == 0:
+            return False
+        worker = workers[0]
+        return worker.num_paused_tasks == num_paused
+
+    threaded_actor = ThreadedActor.options(max_concurrency=actor_concurrency).remote()
+    refs = [  # noqa
+        threaded_actor.main_task.options(name=f"ThreadedActor.main_task_{i}").remote(i)
+        for i in range(20)
+    ]
+
+    wait_for_condition(verify, num_paused=1)
+
+
+@pytest.mark.parametrize("actor_concurrency", [1, 3])
+def test_worker_paused_async_actor(shutdown_only, actor_concurrency):
+    ray.init(_system_config=_SYSTEM_CONFIG)
+
+    @ray.remote
+    class AsyncActor:
+        async def main_task(self, i):
+            if i == 0:
+                import time
+
+                # Pause 5 seconds inside debugger
+                with ray._private.worker.global_worker.worker_paused_by_debugger():
+                    time.sleep(5)
+
+    def verify(num_paused):
+        workers = list_workers(
+            filters=[("num_paused_tasks", "=", num_paused)], detail=True
+        )
+        if len(workers) == 0:
+            return False
+        worker = workers[0]
+        return worker.num_paused_tasks == num_paused
+
+    async_actor = AsyncActor.options(max_concurrency=actor_concurrency).remote()
+    refs = [  # noqa
+        async_actor.main_task.options(name=f"AsyncActor.main_task_{i}").remote(i)
+        for i in range(20)
+    ]
+
+    wait_for_condition(verify, num_paused=1)
```

Comment:
would it be easy to add a test that tests no deadlock when mulitple threads trying to be paused by the debugger thus updating the counter?

Reply:
good idea, I updated the threaded actor to actually paused multiple thread simultaneously and checks they enter and exist the paused state.